### ISSUE TITLE
Type annotations on default function parameters broken

### ISSUE THREAD

**Issue by dharries** [babel/babel#1168]
The code:
```javascript
function test(x: string = "hi"): string {
  return x;
}
```
Results in the following error:
```
SyntaxError: core/test/utils.js: Unexpected token (42:24)
  40 | }
  41 |
> 42 | function test(x: string = "hi"): string {
     |                         ^
  43 |   return x;
  44 | }
  45 |
    at Parser.pp.raise (node_modules/babel/node_modules/babel-core/lib/acorn/src/location.js:68:13)
    at Parser.pp.unexpected (node_modules/babel/node_modules/babel-core/lib/acorn/src/parseutil.js:87:8)
    at Parser.pp.expect (node_modules/babel/node_modules/babel-core/lib/acorn/src/parseutil.js:81:26)
    at Parser.pp.parseBindingList (node_modules/babel/node_modules/babel-core/lib/acorn/src/lval.js:122:40)
    at Parser.pp.parseFunctionParams (node_modules/babel/node_modules/babel-core/lib/acorn/src/statement.js:471:22)
    at Parser.parseFunctionParams (node_modules/babel/node_modules/babel-core/lib/acorn/plugins/flow.js:812:13)
    at Parser.pp.parseFunction (node_modules/babel/node_modules/babel-core/lib/acorn/src/statement.js:464:8)
    at Parser.pp.parseFunctionStatement (node_modules/babel/node_modules/babel-core/lib/acorn/src/statement.js:235:15)
    at Parser.pp.parseStatement (node_modules/babel/node_modules/babel-core/lib/acorn/src/statement.js:67:19)
    at Parser.parseStatement (node_modules/babel/node_modules/babel-core/lib/acorn/plugins/flow.js:631:22)
    at Parser.pp.parseTopLevel (node_modules/babel/node_modules/babel-core/lib/acorn/src/statement.js:22:21)
    at Object.parse (node_modules/babel/node_modules/babel-core/lib/acorn/src/index.js:109:12)
```
I think this is a regression introduced in v5 (my tests broke upgrading 4.7.16 -> 5.0.7).  I managed to get babel happy by swapping the order of the type and default value:
```javascript
function test(x = "hi": string): string
```
But then flow complains about this.

The example above is similar to one from the [flow tests](https://github.com/facebook/flow/blob/master/tests/optional/optional_param.js#L22-L25).  Adding the extra `?`s present in the flow test doesn't seem to help.

**Comment by sebmck** [comment]
Crazy, this is definently a regression. I'm going to blame `esprima-fb` for not having expansive enough tests :stuck_out_tongue_winking_eye:

**Comment by sebmck** [comment]
Fixed as of 5.0.9, thanks!
